runtime/src/iree/base/internal/atomics_msvc.h

```diff
@@ -16,85 +16,273 @@
+  iree_memory_order_release = _Atomic_memory_order_release,
+  iree_memory_order_acq_rel = _Atomic_memory_order_acq_rel,
+  iree_memory_order_seq_cst = _Atomic_memory_order_seq_cst,
+} iree_memory_order_t;
+
+#define IREE_ATOMIC_VAR_INIT(value) (value)
+
+typedef _Atomic int32_t iree_atomic_int32_t;
+typedef _Atomic int64_t iree_atomic_int64_t;
+typedef _Atomic uint32_t iree_atomic_uint32_t;
+typedef _Atomic uint64_t iree_atomic_uint64_t;
+// TODO(#3453): check for __int128 support before using
+// typedef _Atomic __int128 iree_atomic_int128_t;
+typedef _Atomic intptr_t iree_atomic_intptr_t;
+
+#define iree_atomic_thread_fence(order) atomic_thread_fence(order)
+
+#define iree_atomic_load(object, order) __c11_atomic_load((object), (order))
+#define iree_atomic_store(object, desired, order) \
+  __c11_atomic_store((object), (desired), (order))
+#define iree_atomic_fetch_add(object, operand, order) \
+  __c11_atomic_fetch_add((object), (operand), (order))
+#define iree_atomic_fetch_sub(object, operand, order) \
+  __c11_atomic_fetch_sub((object), (operand), (order))
+#define iree_atomic_fetch_and(object, operand, order) \
+  __c11_atomic_fetch_and((object), (operand), (order))
+#define iree_atomic_fetch_or(object, operand, order) \
+  __c11_atomic_fetch_or((object), (operand), (order))
+#define iree_atomic_fetch_xor(object, operand, order) \
+  __c11_atomic_fetch_xor((object), (operand), (order))
+#define iree_atomic_exchange(object, operand, order) \
+  __c11_atomic_exchange((object), (operand), (order))
+#define iree_atomic_compare_exchange_strong(object, expected, desired,  \
+                                            order_succ, order_fail)     \
+  __c11_atomic_compare_exchange_strong((object), (expected), (desired), \
+                                       (order_succ), (order_fail))
+#define iree_atomic_compare_exchange_weak(object, expected, desired,  \
+                                          order_succ, order_fail)     \
+  __c11_atomic_compare_exchange_weak((object), (expected), (desired), \
+                                     (order_succ), (order_fail))
+
+#elif __cplusplus
+
+// When compiling for C++ we reinterpret atomics as std::atomic<T>. This relies
+// on std::atomic on primitive types being lock-free such that the memory for
+// each atomic is just the atomic value. We need this special path because MSVC
+// doesn't support C features like _Generic in C++.
+
+extern "C++" {
+#include <atomic>
+}  // extern "C++"
+
 extern "C" {
-#endif
 
 typedef enum iree_memory_order_e {
-  iree_memory_order_relaxed,
+  iree_memory_order_relaxed = _Atomic_memory_order_relaxed,
+  iree_memory_order_consume = _Atomic_memory_order_consume,
+  iree_memory_order_acquire = _Atomic_memory_order_acquire,
+  iree_memory_order_release = _Atomic_memory_order_release,
+  iree_memory_order_acq_rel = _Atomic_memory_order_acq_rel,
+  iree_memory_order_seq_cst = _Atomic_memory_order_seq_cst,
+} iree_memory_order_t;
```

Comment:
Seeing errors building with this locally:
```
[build] [370/7326   4% :: 21.403] Building CXX object tools\CMakeFiles\iree-benchmark-module.dir\iree-benchmark-module-main.cc.obj
[build] FAILED: tools/CMakeFiles/iree-benchmark-module.dir/iree-benchmark-module-main.cc.obj 
[build] ccache C:\PROGRA~2\MICROS~1\2022\Preview\VC\Tools\MSVC\1431~1.311\bin\Hostx64\x64\cl.exe  /nologo /TP -DBENCHMARK_STATIC_DEFINE -DIREE_HAVE_HAL_EXECUTABLE_EMBEDDED_ELF_PLUGIN=1 -DIREE_HAVE_HAL_EXECUTABLE_LOADER_EMBEDDED_ELF=1 -DIREE_HAVE_HAL_EXECUTABLE_LOADER_SYSTEM_LIBRARY=1 -DIREE_HAVE_HAL_EXECUTABLE_LOADER_VMVX_MODULE=1 -DIREE_HAVE_HAL_EXECUTABLE_SYSTEM_LIBRARY_PLUGIN=1 -DIREE_HAVE_HAL_LOCAL_SYNC_DRIVER_MODULE=1 -DIREE_HAVE_HAL_LOCAL_TASK_DRIVER_MODULE=1 -DIREE_HAVE_HAL_VULKAN_DRIVER_MODULE=1 -DIREE_HAVE_VMVX_MODULE -DIREE_TASK_CPUINFO_DISABLED=1 -DIREE_TRACING_MODE=2 -DIREE_TRACING_PROVIDER_H=\"iree/base/tracing/tracy.h\" -DTRACY_ENABLE -ID:\dev\projects\iree\third_party\benchmark\include -ID:\dev\projects\iree\runtime\src -ID:\dev\projects\iree-build\runtime\src -ID:\dev\projects\iree-build\runtime\src\iree\base\internal\flatcc -ID:\dev\projects\iree-build\runtime\src\iree\schemas -ID:\dev\projects\iree\third_party\vulkan_headers\include -external:ID:\dev\projects\iree -external:ID:\dev\projects\iree-build -external:ID:\dev\projects\iree\third_party\tracy\public -external:ID:\dev\projects\iree\third_party\flatcc\include -external:W0 /DWIN32 /D_WINDOWS /EHsc /Z7 /O2 /Ob1  -std:c++17 -MD /Zc:preprocessor /DWIN32_LEAN_AND_MEAN /DNOMINMAX /D_USE_MATH_DEFINES /D_CRT_SECURE_NO_WARNINGS /D_CRT_SECURE_CPP_OVERLOAD_STANDARD_NAMES /D_SILENCE_NONFLOATING_COMPLEX_DEPRECATION_WARNING /GR- /bigobj /W3 /wd4200 /wd4018 /wd4146 /wd4244 /wd4267 /wd4005 /wd4065 /wd4141 /wd4624 /wd4576 /wd5105 -ID:/dev/projects/iree/third_party/flatcc/include/ -ID:/dev/projects/iree/third_party/flatcc/include/flatcc/reflection/ /showIncludes /Fotools\CMakeFiles\iree-benchmark-module.dir\iree-benchmark-module-main.cc.obj /Fdtools\CMakeFiles\iree-benchmark-module.dir\ /FS -c D:\dev\projects\iree\tools\iree-benchmark-module-main.cc
[build] D:\dev\projects\iree\runtime\src\iree/base/internal/atomics_msvc.h(91): error C2065: '_Atomic_memory_order_relaxed': undeclared identifier
[build] D:\dev\projects\iree\runtime\src\iree/base/internal/atomics_msvc.h(92): error C2065: '_Atomic_memory_order_consume': undeclared identifier
[build] D:\dev\projects\iree\runtime\src\iree/base/internal/atomics_msvc.h(93): error C2065: '_Atomic_memory_order_acquire': undeclared identifier
[build] D:\dev\projects\iree\runtime\src\iree/base/internal/atomics_msvc.h(94): error C2065: '_Atomic_memory_order_release': undeclared identifier
[build] D:\dev\projects\iree\runtime\src\iree/base/internal/atomics_msvc.h(95): error C2065: '_Atomic_memory_order_acq_rel': undeclared identifier
[build] D:\dev\projects\iree\runtime\src\iree/base/internal/atomics_msvc.h(96): error C2065: '_Atomic_memory_order_seq_cst': undeclared identifier
```

Can we fallback to something? I can also update my MSVC
* `VCToolsInstallDir == C:\\Program Files (x86)\\Microsoft Visual Studio\\2022\\Preview\\VC\\Tools\\MSVC\\14.31.31103`
* `UCRTVersion == 10.0.20348.0`

Reply:
Checking if this will help:
![image](https://github.com/user-attachments/assets/0873775f-12f1-4a8e-a1cb-2953bf4a873b)